presto-main/src/main/java/com/facebook/presto/operator/MapBlockEncodingBuffers.java

```diff
@@ -0,0 +1,371 @@
+
+        keyBuffers.accumulateRowSizes(keyOffsetsOffsets, positionCount, rowSizes);
+        valueBuffers.accumulateRowSizes(valueOffsetsOffsets, positionCount, rowSizes);
+    }
+
+    @Override
+    protected void accumulateRowSizes(int[] offsetsOffsets, int positionCount, int[] rowSizes)
+    {
+        // The nested level positionCount could be 0.
+        if (this.positionCount == 0) {
+            return;
+        }
+
+        int lastOffset = offsetsOffsets[0];
+        for (int i = 0; i < positionCount; i++) {
+            int offset = offsetsOffsets[i + 1];
+            rowSizes[i] += POSITION_SIZE * (offset - lastOffset);
+            lastOffset = offset;
+            offsetsOffsets[i + 1] = offsets[offset];
+        }
+
+        // offsetsOffsets might be modified by the next level. Save it for the valueBuffers first.
+        ensureOffsetsOffsetsCapacity(positionCount);
+        System.arraycopy(offsetsOffsets, 0, valueOffsetsOffsets, 0, positionCount + 1);
+
+        keyBuffers.accumulateRowSizes(offsetsOffsets, positionCount, rowSizes);
+        valueBuffers.accumulateRowSizes(offsetsOffsets, positionCount, rowSizes);
+    }
+
+    @Override
+    public void setNextBatch(int positionsOffset, int batchSize)
+    {
+        this.positionsOffset = positionsOffset;
+        this.batchSize = batchSize;
+
+        if (this.positionCount == 0) {
+            return;
+        }
+
+        int beginOffset = offsets[positionsOffset];
+        int endOffset = offsets[positionsOffset + batchSize];
+
+        keyBuffers.setNextBatch(beginOffset, endOffset - beginOffset);
+        valueBuffers.setNextBatch(beginOffset, endOffset - beginOffset);
+    }
+
+    @Override
+    public void copyValues()
+    {
+        if (batchSize == 0) {
+            return;
+        }
+
+        appendNulls();
+        appendOffsets();
+        appendHashTables();
+
+        keyBuffers.copyValues();
+        valueBuffers.copyValues();
+
+        bufferedPositionCount += batchSize;
+    }
+
+    public void serializeTo(SliceOutput output)
+    {
+        writeLengthPrefixedString(output, NAME);
+
+        // Map type
+        TypeSerde.writeType(output, columnarMap.getKeyType());
+
+        keyBuffers.serializeTo(output);
+        valueBuffers.serializeTo(output);
+
+        // Hash tables
+        int hashTableSize = lastOffset * HASH_MULTIPLIER;
+
+        if (hashTableBufferIndex == 0) {
+            output.appendInt(-1);
+        }
+        else {
+            output.appendInt(hashTableSize);
+            output.appendBytes(hashTablesBuffer, 0, hashTableBufferIndex);
+        }
+
+        // positionCount
+        output.writeInt(bufferedPositionCount);
+
+        // offsets
+        output.appendInt(0);
+        if (offsetsBufferIndex > 0) {
+            output.appendBytes(offsetsBuffer, 0, offsetsBufferIndex);
+        }
+
+        serializeNullsTo(output);
+    }
+
+    @Override
+    public long getRetainedSizeInBytes()
```

Comment:
columnarMap.getRetainedSizeInBytes() is missing

Reply:
It should be counted in DecodedBlockNode however the change was lost while separating the commits. I'll add it back there. Counting here in every BlockEncodingBuffers is over counting.

Reply:
@yingsu00 Would you add a comment here explaining why `columnarMap` should not be included? It is not obvious.